src/singlelayerqg.jl

```diff
@@ -0,0 +1,509 @@
+
+Returns the domain-averaged kinetic energy of the fluid, 
+```math
+\\textrm{KE} = \\int \\frac1{2} |\\boldsymbol{\\nabla} \\psi|^2 \\frac{\\mathrm{d}^2 \\boldsymbol{x}}{L_x L_y} = \\sum_{\\boldsymbol{k}} \\frac1{2} |\\boldsymbol{k}|^2 |\\hat{\\psi}|^2 \\ .
+```
+"""
+function kinetic_energy(sol, vars, params, grid)
+  streamfunctionfrompv!(vars.ψh, sol, params, grid)
+  @. vars.uh = sqrt.(grid.Krsq) * vars.ψh      # vars.uh is a dummy variable
+
+  return parsevalsum2(vars.uh , grid) / (2 * grid.Lx * grid.Ly)
+end
+
+kinetic_energy(prob) = kinetic_energy(prob.sol, prob.vars, prob.params, prob.grid)
+
+"""
+    potential_energy(prob)
+    potential_energy(sol, grid, vars, params)
+
+Returns the domain-averaged potential energy of the fluid, 
+```math
+\\textrm{PE} = \\int \\frac1{2} \\frac{\\psi^2}{\\ell^2} \\frac{\\mathrm{d}^2 \\boldsymbol{x}}{L_x L_y} = \\sum_{\\boldsymbol{k}} \\frac1{2} \\frac{|\\hat{\\psi}|^2}{\\ell^2} \\ .
+```
+"""
+function potential_energy(sol, vars, params::EquivalentBarotropicQGParams, grid)
+  streamfunctionfrompv!(vars.ψh, sol, params, grid)
+  return 1 / params.deformation_radius^2 * parsevalsum2(vars.ψh, grid) / (2 * grid.Lx * grid.Ly)
+end
+
+@inline potential_energy(sol, vars, params::BarotropicQGParams, grid) = 0
+
+@inline potential_energy(prob) = potential_energy(prob.sol, prob.vars, prob.params, prob.grid)
+
+"""
+    energy(prob)
+    energy(sol, grid, vars, params)
+
+Returns the domain-averaged total energy of the fluid, that is, the kinetic energy for a
+pure barotropic flow or the sum of kinetic and potential energies for an equivalent barotropic
+flow.
+"""
+@inline energy(prob) = energy(prob.sol, prob.vars, prob.params, prob.grid)
+
+@inline energy(sol, vars, params::BarotropicQGParams, grid) = kinetic_energy(sol, vars, params, grid)
+
+@inline energy(sol, vars, params::EquivalentBarotropicQGParams, grid) = kinetic_energy(sol, vars, params, grid) + potential_energy(sol, vars, params, grid)
+
+"""
+    enstrophy(prob)
+    enstrophy(sol, vars, params, grid)
+
+Returns the domain-averaged enstrophy
+```math
+\\int \\frac1{2} (q + \eta)^2 \\frac{\\mathrm{d}^2 \\boldsymbol{x}}{L_x L_y} = \\sum_{\\boldsymbol{k}} \\frac1{2} |\\hat{q} + \hat{\\eta}|^2 \\ .
+```
+"""
+function enstrophy(sol, vars, params, grid)
+  @. vars.qh = sol
+  return parsevalsum2(vars.qh + params.etah, grid) / (2 * grid.Lx * grid.Ly)
+end
+
+@inline enstrophy(prob) = enstrophy(prob.sol, prob.vars, prob.params, prob.grid)
+
+"""
+    reduced_enstrophy(prob)
+    reduced_enstrophy(sol, vars, params, grid)
+
+Returns the domain-averaged reduced enstrophy ½ ∫(q² + 2qη) dxdy / (Lx Ly) .
+```math
+\\int \\frac1{2} (q^2 + 2 q \eta) \\frac{\\mathrm{d}^2 \\boldsymbol{x}}{L_x L_y} = \\sum_{\\boldsymbol{k}} \\frac1{2} |\\hat{q}|^2 + 2 \hat{q} \hat{\eta} \\ .
+```
+"""
+function reduced_enstrophy(sol, vars, params, grid)
+  @. vars.qh = sol
+  return parsevalsum(abs2.(vars.qh) .+ 2 * vars.qh .* params.etah, grid) / (2 * grid.Lx * grid.Ly)
+end
```

Comment:
@liasiegelman,

1) what is this quantity "reduced enstrophy"? I've never encountered it elsewhere. (I'm just wondering if it's not something of general interest then perhaps we shouldn't include it in the module but the users can define it as a custom diagnostic?)

2) I'm pretty sure there is a `conj()` missing from the definition? I would expect that `∫ a b dx = ∑ conj(â)*b̂`...

Reply:
Both `enstrophy` and `reduced_enstrophy` only differ by a constant, correct?
```julia
reduced_enstrophy = enstrophy - ∫ η² d²x / (Lx * Ly)
```

Reply:
@navidcy Hi Navid, you are right. It's just a name we came up with Bill to look at the variable part of the enstrophy. It should probably not be implemented in the module.

Reply:
354d613